### ISSUE TITLE
Reproduction issues on the MatBench dataset

### ISSUE THREAD

**Issue by Wai-BiBaBu** [Bin-Cao/PRDNet#1]
Hi, this is an impressive work, and thank you for providing the code base. 

By introducing diffraction information in reciprocal space based on pseudo-particles, PRDNet demonstrates outstanding performance across multiple benchmarks. Its performance on **MatBench** is particularly remarkable. However, as shown in the table below, I **observed performance discrepancies while attempting to reproduce the PRDNet results on MatBench** and sincerely hope the authors can provide guidance.

The reproduced PRDNet's MAE in some datasets is close to that of the classical CGCNN model which appears unexpected. When using reciprocal distance similar to iComformer, performance improves but still falls short of what is described in the paper. By the way, the reproduction results of iComformer are not bad, so it probably isn't an issue with dataset processing.

<img width="1706" height="383" alt="Image" src="https://github.com/user-attachments/assets/6d28083b-dc6e-4c26-9052-e716ab3905d5" />

The details of the reproduction are as follows: 

- **Dataset Processing**: Following the requirements of MatBench, a five-fold split was performed using the random seed 18012019. Then, using the same random seed (18012019), the non-test set was divided into training and validation sets in a 90%:10% ratio. The **12-nearest neighbors** approach from Matformer was used (neighbors with distances less than that to the 12th nearest neighbor were considered bonded). The `data.pos` passed to the diffraction integration module (`self.diffraction_integration`) corresponds to the **fractional coordinates** of atoms. Node features are represented by CGCNN's 92-dimensional features, and edge features are Euclidean distances.

- **Model Hyperparameters**: Consistent with **PrdnetConfig's default hyperparameters**. **Due to encountering errors, two modifications were made**: changing the input dimension of the first fc_layers layer (`model.py:184`) from `config.node_features` (256) to `config.fc_features` (512); modifying `graph_features` in `self.diffraction_integration` (`model.py:217`) from `config.fc_features` (512) to `config.node_features` (256).

- **Training Hyperparameters**: Aligned with main training configurations from Matformer and iComformer, using a **onecycle** scheduler with pct_start=0.3 and **AdamW** optimizer trained for **500 epochs** at a learning rate of **0.0005** (0.0001 for iComformer and PRDNet using reciprocal distance, but it should not affect the conclusion.) with weight decay set at **1e-5** and batch size fixed at **64**; MSE served as loss function during optimization process. The model metrics displayed in the table are obtained by calculating **the average MAE across five folds**. 

Considering that the main repository code was uploaded approximately five to six months ago, could you please update the code required to reproduce the results reported in the paper? Providing more detailed hyperparameter specifications would also be highly beneficial. Thanks a lot.

**Comment by Bin-Cao** [comment]
Thank you very much for your interest in our work and for taking the time to carefully reproduce the results. We sincerely appreciate the detailed description of your experimental setup.

Regarding the discrepancy you observed on the MatBench datasets, several factors may contribute:

**Graph construction and neighbor selection.** In the original experiments, the graph construction strategy and neighbor cutoff differ from the 12-nearest-neighbor approach used in Matformer. In PRDNet, we used 18 nearest neighbors.

**Feature configuration.** In our implementation, the node and graph feature dimensions are tightly coupled with the diffraction integration module. Changing the input dimensions of `fc_layers` and `graph_features`, as described, may affect the information flow in the model.

**Training configuration.** Although the overall setup is similar, PRDNet can be somewhat sensitive to the learning rate schedule and other training configurations. The settings used in the paper were slightly tuned to ensure stable training.

I will also check whether there are any differences in the specific settings used in our original experiments when my schedule allows, and I will follow up with you if I find anything relevant.

Thanks again for your careful reproduction and for bringing this to our attention.

If you are also interested, you may try reproducing the results on the MP dataset for additional verification.

**Comment by Wai-BiBaBu** [comment]
Thank you very much for taking the time to address our concerns and for providing such detailed guidance. Following your suggestions, we have conducted further experiments, but still encountered some difficulties.

- We adjusted the number of nearest neighbors used in constructing the crystal graph **from 12 to 18**. However, the table below demonstrates that **altering the number of nearest neighbors yields only marginal fluctuations in MAE**, implying that it may not be the main bottleneck responsible for the model's poor performance in our reproduction.

<img width="2476" height="423" alt="Image" src="https://github.com/user-attachments/assets/cea1ea18-83fc-4c0b-b0f7-4078a42fe2cb" />


- **About feature configuration**
We fully agree with your insight that feature dimensions critically influence information flow and, consequently, model performance. However, we wish to clarify that the modifications made to `self.diffraction_integration` (`model.py:217`) and the input dimension of the first fc_layers (`model.py:184`) were **made out of necessity**. **These adjustments were necessitated by dimension mismatch errors encountered.** For example, when the diffraction module is enabled, the `features` tensor (`model.py:297`) has a dimension equal to `config.fc_features=512` (as defined in line 218: `output_features=config.fc_features`). Consequently, keeping the input dimension of `fc_layers` at the source code's default of `config.node_features=256` would result in a runtime error due to incompatible shapes. Therefore, to resolve this incompatibility, we changed the input dimension of the first fc_layers (`model.py:184`) from `config.node_features (256)` to `config.fc_features (512)`.
Additionally, we noticed two locations in `diffraction.py` where dimensionality reductions may introduce **low-rank bottlenecks**. Specifically, lines 113–122 compress features from 256 down to 128 before expanding to 300 (`self.num_hkl`), while lines 199–208 reduce dimensions from 600 (2 * `num_hkl`) down to 256 before projecting back to 512 (`graph_features`). These narrow intermediate layers may limit the model's representational capacity. Consequently, we are also curious whether the code in these sections has been altered or optimized in the version used for the reported results.

Thank you for offering to check whether there are any differences in the specific settings used in your original experiments. We look forward to your follow-up if anything relevant is found. In addition, it would be especially helpful if you could pay particular attention to the issues about feature configuration mentioned above.

We are also very interested in applying PRDNet to datasets such as MP and Jarvis; however, their greater number of structures leads to substantially higher training costs compared to MatBench. Moreover, MatBench employs five-fold cross-validation, whereas MP and Jarvis report MAE from a single run—making MatBench results less susceptible to random variation. We would prefer to first resolve the MatBench reproduction issues before moving forward. (After all, our current MAE on `log_gvrh` and `dielectric` is **47%** and **85%** higher, respectively, than the values reported in the paper—**suggesting that, aside from hyperparameter tuning of training configuration, other issues may also be at play.**)

**Comment by Bin-Cao** [comment]
Thank you very much for sharing your updated results. To better understand the differences, could you please share your code with me? My team and I will review it carefully and align the configuration and data processing. I will also have my student assist in checking the details.

Regarding the results reported in the paper, the baseline values for MatBench were recorded by our group. We want to make sure the settings are properly aligned. The model has several historical versions, but I believe the differences are not due to hyperparameter settings in message passing; the reason may be some variations in the data embedding layers.

Please feel free to share the code whenever convenient. You can reach me at `bcao686@connect.hkust-gz.edu.cn`.

Thank you again for your support and professionalism. We also welcome any suggestions to `improve the strategy`  and hope to collaborate further. Since we have seen that the diffraction layer consistently enhances performance, we are `excited about potential joint work`.

**Comment by Bin-Cao** [comment]
good work many thx!
